### PR TITLE
feat: support equality_delta_writer

### PR DESCRIPTION
**Pull request by ZENOTME**

### CODE REVIEW

File: icelake/src/io/file_writer/equality_delta_writer.rs [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+            )
+            .await?,
+            inserted_rows: HashMap::new(),
+            row_converter,
+            primary_column_idx,
+        })
+    }
+
+    /// Write the batch.
+    pub async fn write(&mut self, batch: RecordBatch) -> Result<()> {
```
Comment (liurenjie1024) [inline icelake/src/io/file_writer/equality_delta_writer.rs:137]: The approach of separating `write` and `delete` maybe not quite convenient. Thinking about how risingwave will use it: the sink writer will accept a `StreamChunk`, it needs to call `write/delete` row by row. How about providing a method called `delte_write` like this:
```rust
pub async fn delta_write(&mut self, ops: IntArray, batch: RecordBatch) -> Result<()> {}
```

Where the ops array is for insert/delete operator, we can use 1 for insertion, 2 for deletion.

File: icelake/tests/delta_test.rs [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+    docker_compose.run();
+
+    DeltaTest::new(docker_compose, poetry, catalog)
+}
+
+fn main() {
+    // Parse command line arguments
+    let args = Arguments::from_args();
+
+    let catalogs = vec!["storage"];
```
Comment (liurenjie1024) [inline icelake/tests/delta_test.rs:403]: Missing rest catalog?

File: icelake/tests/delta_test.rs [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+                "-q2",
+                "select * from s1.t2",
+            ],
+            "Check select * from s1.t1",
+        )
+    }
+
+    async fn run_equality_delta_write_test(mut self) {
+        self.init();
+        self.test_delete().await;
```
Comment (liurenjie1024) [inline icelake/tests/delta_test.rs:360]:
```suggestion
        self.test_write().await;
```

File: icelake/tests/delta_test.rs [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+            .await
+            .unwrap()
+            .build_equality_delta_writer(vec![1, 2])
+            .await
+            .unwrap();
+
+        self.write_and_delete_with_delta(
+            &table,
+            &mut delta_writer,
+            Some(vec![
```
Comment (liurenjie1024) [inline icelake/tests/delta_test.rs:329]: I don't quite understand why the results are same? I think the result should be:
```
1 1 1
1 2 4
1 2 4
```

Comment (ZENOTME) [reply icelake/tests/delta_test.rs:329]: We first insert 
```
row: 
1 1 1 
1 2 2 
1 4 4 
```
and then insert 
```
row: 
1 3 3 
1 5 5
```
and remove
```
row: 
1 4 4 
1 5 5
```

File: icelake/src/io/file_writer/equality_delta_writer.rs [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+            }
+        }
+
+        self.data_file_writer.write(batch).await?;
+
+        Ok(())
+    }
+
+    /// Delete the batch.
+    pub async fn delete(&mut self, batch: RecordBatch) -> Result<()> {
```
Comment (liurenjie1024) [inline icelake/src/io/file_writer/equality_delta_writer.rs:211]: Same question as above?

File: icelake/src/io/file_writer/equality_delta_writer.rs [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Write the batch.
+    /// 1. If a row wtih the same unique column is not written, then insert it.
+    /// 2. If a row with the same unique column is written, then delete the previous row and insert the new row.
+    pub async fn write(&mut self, batch: RecordBatch) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (ZENOTME) [reply icelake/src/io/file_writer/equality_delta_writer.rs:186]: I think exposing these interface is ok.🤔

[the remark this answers is not in the document]
Comment (Xuanwo) [reply icelake/src/io/file_writer/equality_delta_writer.rs:186]: Will we allow users to operate on those writers directly?

[the remark this answers is not in the document]
Comment (liurenjie1024) [reply icelake/src/io/file_writer/equality_delta_writer.rs:186]: It gives more flexibility for users. OK to me.

[the remark this answers is not in the document]
Comment (ZENOTME) [reply icelake/src/io/file_writer/equality_delta_writer.rs:186]: > Will we allow users to operate on those writers directly?

Would you mean write interface or writer in file_writer.

[the remark this answers is not in the document]
Comment (Xuanwo) [reply icelake/src/io/file_writer/equality_delta_writer.rs:186]: > writer in file_writer.

The `write` API in file_writer.

**Pull request merged**